pulp_ansible/app/galaxy/v3/views.py

```diff
@@ -144,45 +145,66 @@ def filter_queryset(self, queryset):
             else:
                 raise ValidationError("Cannot parse value of `deprecated` GET parameter")
 
-        return queryset
+        return super().filter_queryset(queryset)
 
     def get_queryset(self):
         """
         Returns a Collections queryset for specified distribution.
         """
         repo_version = self.get_repository_version(self.kwargs["path"])
+
+        versions = CollectionVersion.objects.filter(
+            version_memberships__repository=repo_version.repository
+        ).annotate(
+            repo_version_added_at=Window(
+                expression=FirstValue("version_memberships__pulp_last_updated"),
+                partition_by=[F("collection_id")],
+                order_by=F("version_memberships__pulp_last_updated").desc(),
+            ),
+            repo_version_removed_at=Window(
+                expression=FirstValue("version_memberships__version_removed__pulp_last_updated"),
+                partition_by=[F("collection_id")],
+                order_by=F("version_memberships__version_removed__pulp_last_updated").desc(
+                    nulls_last=True
+                ),
+            ),
+        )
         deprecated_query = AnsibleCollectionDeprecated.objects.filter(
-            collection=OuterRef("pk"), repository_version=repo_version
+            collection=OuterRef("pk"),
+            repository_version=repo_version,
+        ).only("collection_id")
+
+        collections_qs = Collection.objects.filter(
```

Comment:
This is definitely better for now at least for AH production environment, since the number of collections is relatively low. However I would prefer limiting the number of fetched rows to list of collections that are returned. Which means this query could be applied after pagination with a filter by collection ids, returned by collections queryset